databricks/koalas/indexes.py

```diff
@@ -154,14 +156,50 @@ def _with_new_scol(self, scol: spark.Column) -> "Index":
         :param scol: the new Spark Column
         :return: the copied Index
         """
-        sdf = self._internal.spark_frame.select(scol.alias(SPARK_DEFAULT_INDEX_NAME))
-        internal = InternalFrame(
-            spark_frame=sdf,
-            index_spark_columns=[scol_for(sdf, col) for col in sdf.columns],
-            index_names=self._internal.index_names,
+        internal = self._internal.copy(
+            index_spark_columns=[scol.alias(SPARK_DEFAULT_INDEX_NAME)],
+            column_labels=[],
+            data_spark_columns=[],
         )
         return DataFrame(internal).index
 
+    def _need_alignment_for_column_op(self, other: "IndexOpsMixin") -> bool:
+        return self._internal.spark_frame is not other._internal.spark_frame
+
+    def _align_and_column_op(self, f, *args) -> "Index":
+        if get_option("compute.ops_on_diff_frames"):
+            # This could cause as many counts, reset_index calls, joins for combining
+            # as the number of `Index`s in `args`. So far it's fine since we can assume the ops
+            # only work between at most two `Index`s. We might need to fix it in the future.
+            self_len = len(self)
+            if any(len(col) != self_len for col in args if isinstance(col, IndexOpsMixin)):
+                raise ValueError("operands could not be broadcast together with shapes")
```

Comment:
How does length comparison relate to broadcast in the error message?

Reply:
The term "broadcast" here is not the same as Spark's "broadcast".
Maybe the term should've been "broad-cast", which means cast to broader types, or in this case cast to larger size.

Reply:
I see, maybe that's what we are talking about
```
>>> z= np.arange(12).reshape(3,4)
>>> m= np.arange(9).reshape(3,3)
>>> z * m
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    z * m
ValueError: operands could not be broadcast together with shapes (3,4) (3,3)
```

The current error message also looks good.